messaging-activemq/src/main/asciidoc/specs/WFLY-5838_artemis_thread_pools.adoc

```diff
@@ -0,0 +1,440 @@
+
+Relevant methods are:
+
+----
+* ServiceRegistry#setExecutorService(ExecutorService)
+* ServiceRegistry#setIOExecutorService(ExecutorService)
+* ServiceRegistry#setScheduledExecutorService(ScheduledExecutorService)
+----
+
+For global thread pools, the  WildFly `messaging-activemq` subsystem must call https://github.com/apache/activemq-artemis/blob/master/artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/client/ActiveMQClient.java#L197[ActiveMQClient#injectPools] __before any Artemis client or server is started__:
+
+----
+ActiveMQClient#injectPools(ExecutorService, ScheduledExecutorService)
+----
+
+WARNING: Artemis requires that *both the regular and scheduled thread pools are injected* (it is not possible to inject only one and not the other). This must be enforced by WildFly `messaging-activemq` subsystem model.
+
+== Uninjectable Thread Pools in Artemis
+
+There are still a few places where Artemis creates thread pools and do not allow to inject them from WildFly.
+
+* https://github.com/apache/activemq-artemis/blob/master/artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/impl/invm/InVMConnector.java#L109[InVMConnector#getInVMExecutor]
+** At first glance this ExecutorService should be replaced by the actual global client thread pool returned by `ActiveMQClient#getGlobalThreadPool()` (which is injectable). If the Artemis development team objects that each InVMConnector should have its own unique thread pool, they will have to provide a way to inject it.
+
+CAUTION: We need feedback from the Artemis team on this item.
+
+* https://github.com/apache/activemq-artemis/blob/master/artemis-journal/src/main/java/org/apache/activemq/artemis/core/journal/impl/JournalImpl.java#L2316[JournalImpl#start]
+** When the Journal is started, it creates a thread pool if the `providedIOThreadPool` parameter is `null`. This code should never be reached as the `providedIOThreadPool` corresponds to the ActiveMQServerImpl's `ioExecutorFactory` that is *always* set in `ActiveMQServerImpl#initializeExecutorServices`.
+
+NOTE: We need feedback from Artemis team but it seems to be some dead code from previous version that has no impact on the actual code path.
+
+* https://github.com/apache/activemq-artemis/blob/master/artemis-core-client/src/main/java/org/apache/activemq/artemis/core/client/impl/ServerLocatorImpl.java#L222[ServerLocatorImpl.java#setThreadPools]
+** In WildFly, this corresponds to any lookup of a JMS ConnectionFactory or use of a Message-Driven Bean. By default, Artemis is configured to use the global client pools. This is configuration using the `connection-factory` and `pooled-connection-factory`'s `use-global-pools` (which is `true` by default).
+If it is set to `false`, Artemis will create separate thread pools for each JMS connection factory or resource adapter.
```

Comment:
What is the intention with these thread pools?

Reply:
@MartinStyk they are used by the connection factory's internal ClientSessionFactory to execute some of its code

Reply:
@jmesnil , do you plan to make it configurable/injectable somehow? It is probably not possible to inject a wildfly thread pools here, but do you consider providing of ThreadFactory from Wildfly?